### PR TITLE
Implement prometheus agent sharding

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/giantswarm/issues/26987

## Checklist

I have:

- [x] Described why this change is being introduced
- [x] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

**Comment by QuentinBisson** [comment]
Code was tested thanks to the cilium testing clusters:

```
 k get cm --context gs-gauss -n v5wq7                               v5wq7-remote-write-config  -oyaml
apiVersion: v1
data:
  values: |
    global: {}
    prometheus-agent:
      externalLabels:
        cluster_id: v5wq7
        cluster_type: workload_cluster
        customer: giantswarm
        installation: gauss
        organization: giantswarm
        pipeline: testing
        provider: aws
        region: eu-west-1
        service_priority: highest
      image:
        tag: v2.43.0-stringlabels
      shards: 1
      version: v2.43.0-stringlabels
kind: ConfigMap
metadata:
  creationTimestamp: "2023-05-17T11:45:51Z"
  name: v5wq7-remote-write-config
  namespace: v5wq7
  resourceVersion: "1216938337"
  uid: 6a2437c4-66e3-46b3-a855-00d2607002a9
> k get cm --context gs-gauss -n wx88s wx88s-remote-write-config
NAME                        DATA   AGE
wx88s-remote-write-config   1      8d
> k get cm --context gs-gauss -n wx88s wx88s-remote-write-config -oyaml
apiVersion: v1
data:
  values: |
    global: {}
    prometheus-agent:
      externalLabels:
        cluster_id: wx88s
        cluster_type: workload_cluster
        customer: giantswarm
        installation: gauss
        organization: marie
        pipeline: testing
        provider: aws
        region: eu-west-1
        service_priority: highest
      image:
        tag: v2.43.0-stringlabels
      shards: 3
      version: v2.43.0-stringlabels
kind: ConfigMap
metadata:
  creationTimestamp: "2023-05-09T10:14:20Z"
  name: wx88s-remote-write-config
  namespace: wx88s
  resourceVersion: "1216775684"
  uid: 8401027d-368f-4e7d-928e-50d881ea8f52

```

**Comment by QuentinBisson** [comment]
We cumulate both because we do not know if a provider will use one, the other or both :D

**Comment by QuentinBisson** [comment]
There is probably an issue on golem :)

File: service/controller/resource/monitoring/remotewriteconfig/shards_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		t.Errorf(`expected computeShards(2, 16) to be %d, got %d`, expected, result)
+	}
+
+	// threshold hit
+	expected = 1
+	result = computeShards(2, 15)
+	if result != expected {
+		t.Errorf(`expected computeShards(2, 15) to be %d, got %d`, expected, result)
+	}
+}
```
Comment (marieroque) [inline service/controller/resource/monitoring/remotewriteconfig/shards_test.go:58]: Can you add an unit test for ScaleUp ?

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteconfig/shards_test.go:58]: I did that's the first test :)

**Comment by QuentinBisson** [comment]
Question asked here about golem:

https://gigantic.slack.com/archives/C03LV8E0RL7/p1684331620179169

I'm wondering if that's not the bastion

**Comment by marieroque** [comment]
> 

Probably, because the name of the machinedeployments is `rp2y6-bastion`

File: service/controller/resource/monitoring/remotewriteconfig/shards_test.go (hunk 1 above)
Comment (marieroque) [reply service/controller/resource/monitoring/remotewriteconfig/shards_test.go:58]: Well yes but that is tested the first step of 20, not the other steps

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteconfig/shards_test.go:64]: I added a test with the next step

**Comment by QuentinBisson** [comment]
@marieroque I added this test to ignore bastion machines https://github.com/giantswarm/prometheus-meta-operator/pull/1296/commits/ee3e4fa35965ed717ac975f9191e0f19bbef387f

**Comment by QuentinBisson** [comment]
I would need to test it on golem

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Dynamically compute Shards according to cluster size.
```
Comment (hervenicol) [inline CHANGELOG.md:12]:
```suggestion
- Dynamically compute agent shards number according to cluster size.
```

**Review by hervenicol — APPROVED** [review]
I like it: clear and easy to read, and unit tests :heart:

**Comment by QuentinBisson** [comment]
Waiting before alerting is done before releasing

**Pull request merged**